src/Yi/Braid.hs

```diff
@@ -0,0 +1,596 @@
+{-# language BangPatterns #-}
+{-# language DeriveDataTypeable #-}
+{-# language LambdaCase #-}
+{-# language MultiParamTypeClasses #-}
+{-# language OverloadedStrings #-}
+{-# language ScopedTypeVariables #-}
+{-# language ViewPatterns #-}
+{-# language FlexibleContexts #-}
+{-# language UndecidableInstances #-}
+{-# language ConstraintKinds #-}
+{-# options_haddock show-extensions #-}
+
+-- |
+-- Module      :  Yi.Braid
+-- License     :  GPL-2
+-- Maintainer  :  yi-devel@googlegroups.com
+-- Stability   :  experimental
+-- Portability :  portable
+--
+-- This module defines a @rope@ data structure for use in Yi. This
+-- specific implementation uses a fingertree over Text.
+--
+-- In contrast to our old implementation, we can now reap all the
+-- benefits of Text: automatic unicode handling and blazing fast
+-- implementation on underlying strings. This frees us from a lot of
+-- book-keeping. We don't lose out on not using ByteString directly
+-- because the old implementation encoded it into UTF8 anyway, making
+-- it unsuitable for storing anything but text.
+
+module Yi.Braid
+  ( Braid(..)
+  , Chunk(..)
+  , HasSize(..)
+  , Yi.Braid.mkChunk
+  , Yi.Braid.overChunk
+  , (Yi.Braid.-|)
+  , (Yi.Braid.|-)
+  , Yi.Braid.reverse
+  , Yi.Braid.toReversed
+  , Yi.Braid.toBraid
+  , Yi.Braid.toBraid'
+  , Yi.Braid.extractBraid
+  , Yi.Braid.null
+  , Yi.Braid.empty
+  , Yi.Braid.length
+  , Yi.Braid.append
+  , Yi.Braid.concat
+  , Yi.Braid.head
+  , Yi.Braid.last
+  , Yi.Braid.init
+  , Yi.Braid.tail
+  , Yi.Braid.splitAt
+  , Yi.Braid.take
+  , Yi.Braid.drop
+  , Yi.Braid.dropWhile
+  , Yi.Braid.dropWhileEnd
+  , Yi.Braid.takeWhile
+  , Yi.Braid.takeWhileEnd
+  , Yi.Braid.span
+  , Yi.Braid.break
+  , Yi.Braid.intercalate
+  , Yi.Braid.intersperse
+  , Yi.Braid.cons
+  , Yi.Braid.snoc
+  , Yi.Braid.singleton
+  , Yi.Braid.any
+  , Yi.Braid.all
+  , Yi.Braid.filter
+  , Yi.Braid.map
+  , Yi.Braid.split
+  , Yi.Braid.withChunk
+  , Yi.Braid.unsafeWithChunk
+  , Yi.Braid.foldl'
+  , Yi.Braid.replicate
+  , Yi.Braid.replicateSegment
+  ) where
+
+import           Control.DeepSeq
+import qualified Data.FingerTree as T
+import           Data.FingerTree hiding (null, empty, reverse, split)
+import qualified Data.List as L (foldl')
+import           Data.Maybe
+import           Data.Monoid
+import           Data.Typeable
+
+import qualified Yi.Segment as S
+
+type ValidBraid v a = (T.Measured v (Chunk a), S.Segmented a, HasSize v)
```

Comment:
I took me a minute to realize this has kind Constraint, so could you add a kind signature?